docs: add `why` section on the top to clarify (#323)
un-ts/eslint-plugin-import-x: README.md

Reviewer: Ah I've been swamped this week but hoped to get to reviewing this before merge. Filing a bookmark to come back to you with a review soon. Sorry for the delay!

This plugin intends to support linting of ES2015+ (ES6+) import/export syntax, and prevent issues with misspelling of file paths and import names. All the goodness that the ES2015+ static module syntax intends to provide, marked up in your editor.

> It started as a fork of [`eslint-plugin-import`] using [`get-tsconfig`] to replace [`tsconfig-paths`] and heavy [`typescript`] under the hood, making it faster, through less heavy dependency on Typescript, and cleaner dependencies altogether.

Reviewer 2: I don't understand. `eslint-plugin-import` only has Typescript in devDependencies same as this library: https://github.com/import-js/eslint-plugin-import/blob/da5f6ec13160cb288338db0c2a00c34b2d932f0d/package.json#L111

Author: Nope, https://github.com/import-js/eslint-plugin-import/blob/da5f6ec13160cb288338db0c2a00c34b2d932f0d/src/exportMap/typescript.js#L16

> It started as a fork of [`eslint-plugin-import`] using [`get-tsconfig`] to replace [`tsconfig-paths`] and heavy [`typescript`] under the hood, making it faster, through less [heavy dependency on Typescript](https://github.com/import-js/eslint-plugin-import/blob/da5f6ec13160cb288338db0c2a00c34b2d932f0d/src/exportMap/typescript.js#L16), and cleaner dependencies altogether.

Author: > How?

replace `tsconfig-paths` and heavy `typescript` with single [`get-tsconfig`](https://github.com/privatenumber/get-tsconfig)?

> Why?

`typescript` is heavy to load at https://github.com/import-js/eslint-plugin-import/blob/da5f6ec13160cb288338db0c2a00c34b2d932f0d/src/exportMap/typescript.js#L16

- [Changelog](#changelog)
- [License](#license)

## Why

Many issues cannot be fixed easily without API changes. For example, see:

- <https://github.com/import-js/eslint-plugin-import/issues/1479>
- <https://github.com/import-js/eslint-plugin-import/issues/2108>
- <https://github.com/import-js/eslint-plugin-import/issues/2111>

[`eslint-plugin-import`] refused to accept BREAKING CHANGES for these issues, so we had to fork it.

[`eslint-plugin-import`] now claims in <https://github.com/un-ts/eslint-plugin-import-x/issues/170> that it will accept BREAKING CHANGES. However, still nothing is happening: <https://github.com/import-js/eslint-plugin-import/pull/3091>.

> We haven't resolved all the issues yet, but we are working on them, which could happen in the next major version (v5): <https://github.com/un-ts/eslint-plugin-import-x/issues/235>.

Reviewer 3: Also mentioning that ljharb simply refuses to implement `exports` feature in the node resolver and even locks the feature request issue https://github.com/import-js/eslint-plugin-import/issues/1810, and on the other hand, `eslint-plugin-import-x` now provides first-party support for `exports` and will become default in the next major version.

Author: @SukkaW Would you like to provide a diff suggestion instead?

Reviewer 3: ```suggestion
[`eslint-plugin-import`] refuses to support the `exports` feature, and the maintainer even locked the feature request issue <https://github.com/import-js/eslint-plugin-import/issues/1810> to prevent future discussion. In the meantime, `eslint-plugin-import-x` now provides first-party support for the `exports` feature <https://github.com/un-ts/eslint-plugin-import-x/pull/209>, which will become the default in the next major version (v5).

We haven't resolved all the issues yet, but we are working on them, which could happen in the next major version (v5): <https://github.com/un-ts/eslint-plugin-import-x/issues/235>.
```

## Differences

So what are the differences from `eslint-plugin-import` exactly?

- we target [Node `^18.18.0 || ^20.9.0 || >=21.1.0`](https://github.com/un-ts/eslint-plugin-import-x/blob/8b2d6d3b612eb57fb68c3fddec25b02fc622df7c/package.json#L12) + [ESLint `^8.57.0 || ^9.0.0`](https://github.com/un-ts/eslint-plugin-import-x/blob/8b2d6d3b612eb57fb68c3fddec25b02fc622df7c/package.json#L71), while `eslint-plugin-import` targets [Node `>=4`](https://github.com/import-js/eslint-plugin-import/blob/da5f6ec13160cb288338db0c2a00c34b2d932f0d/package.json#L6) and [ESLint `^2 || ^3 || ^4 || ^5 || ^6 || ^7.2.0 || ^8 || ^9`](https://github.com/import-js/eslint-plugin-import/blob/da5f6ec13160cb288338db0c2a00c34b2d932f0d/package.json#L115C16-L115C64)
- we don't depend on old and outdated dependencies, so [we have 49 dependencies](https://npmgraph.js.org/?q=eslint-plugin-import-x) compared to [117 dependencies for `eslint-plugin-import`](https://npmgraph.js.org/?q=eslint-plugin-import)
- `eslint-plugin-import` uses `tsconfig-paths` + `typescript` itself to load `tsconfig`s while we use the single `get-tsconfig` instead, which is much faster and cleaner

> - `eslint-plugin-import` uses [`resolve`] which doesn't support the `exports` field in `package.json` while we use [`unrs-resolver`] instead, which does support it and is a [`napi-rs`] project so is also much faster

Reviewer 3: ```suggestion
- `eslint-plugin-import` uses [`resolve`] which doesn't support the `exports` field in `package.json` while we build our own rust-based resolver [`unrs-resolver`] instead, which is feature-rich and way more performant.
```

- [Changelog](#changelog)
- [License](#license)

## Why

> Because a lot of issues can not be fixed easily without API changes, see
> 
> - <https://github.com/import-js/eslint-plugin-import/issues/1479>
> - <https://github.com/import-js/eslint-plugin-import/issues/2108>
> - <https://github.com/import-js/eslint-plugin-import/issues/2111>
> 
> But [`eslint-plugin-import`] refused to accept BREAKING CHANGES for these issues, so we have to fork it.
> 
> And then [`eslint-plugin-import`] claimed to be allowed to update with BREAKING CHANGES at <https://github.com/un-ts/eslint-plugin-import-x/issues/170>,
> but it still doesn't happen or even seems to happen: <https://github.com/import-js/eslint-plugin-import/pull/3091>.
> 
> We haven't resolved all the issues yet, but we are working on it which could happen in the next major version (v5): <https://github.com/un-ts/eslint-plugin-import-x/issues/235>.

Reviewer 2: ```suggestion
A lot of issues can not be fixed easily without API changes. E.g. see:

- <https://github.com/import-js/eslint-plugin-import/issues/1479>
- <https://github.com/import-js/eslint-plugin-import/issues/2108>
- <https://github.com/import-js/eslint-plugin-import/issues/2111>

[`eslint-plugin-import`] refused to accept BREAKING CHANGES for these issues, so we had to fork it.

[`eslint-plugin-import`] now claims in <https://github.com/un-ts/eslint-plugin-import-x/issues/170> that it will accept BREAKING CHANGES. However, still nothing is happening: <https://github.com/import-js/eslint-plugin-import/pull/3091>.

We haven't resolved all the issues yet, but we are working on it which could happen in the next major version (v5): <https://github.com/un-ts/eslint-plugin-import-x/issues/235>.
```

## Differences

> So what's the differences from `eslint-plugin-import` exactly?

Reviewer 2: ```suggestion
So what are the differences from `eslint-plugin-import` exactly?
```

- we target [Node `^18.18.0 || ^20.9.0 || >=21.1.0`](https://github.com/un-ts/eslint-plugin-import-x/blob/8b2d6d3b612eb57fb68c3fddec25b02fc622df7c/package.json#L12) + [ESLint `^8.57.0 || ^9.0.0`](https://github.com/un-ts/eslint-plugin-import-x/blob/8b2d6d3b612eb57fb68c3fddec25b02fc622df7c/package.json#L71), while `eslint-plugin-import` targets [Node `>=4`](https://github.com/import-js/eslint-plugin-import/blob/da5f6ec13160cb288338db0c2a00c34b2d932f0d/package.json#L6) and [ESLint `^2 || ^3 || ^4 || ^5 || ^6 || ^7.2.0 || ^8 || ^9`](https://github.com/import-js/eslint-plugin-import/blob/da5f6ec13160cb288338db0c2a00c34b2d932f0d/package.json#L115C16-L115C64)

> - we don't depend on old and outdated dependencies, so we have short dependency depths, see also <https://github.com/un-ts/eslint-plugin-import-x/pull/11>

Reviewer 2: ```suggestion
- we don't depend on old and outdated dependencies, so [we have 49 dependencies](https://npmgraph.js.org/?q=eslint-plugin-import-x) compared to [117 dependencies for `eslint-plugin-import`](https://npmgraph.js.org/?q=eslint-plugin-import)
```

- `eslint-plugin-import` uses `tsconfig-paths` + `typescript` itself to load `tsconfig`s while we use the single `get-tsconfig` instead, which is much faster and cleaner

> - `eslint-plugin-import` uses [`resolve`] which doesn't support `exports` field in `package.json` while we use [`unrs-resolver`] instead, which is a [napi-rs] project so it's much faster

Reviewer 2: ```suggestion
- `eslint-plugin-import` uses [`resolve`] which doesn't support the `exports` field in `package.json` while we use [`unrs-resolver`] instead, which does support it and is a [`napi-rs`] project so is also much faster
```

> - Our [v3 resolver](./resolvers/README.md#v3) interface shares a single `resolver` instance by default which is used all across resolving chains so it would benefit from caching and memoization out of box

Reviewer 2: ```suggestion
- Our [v3 resolver](./resolvers/README.md#v3) interface shares a single `resolver` instance by default which is used all across resolving chains so it would benefit from caching and memoization out-of-the-box
```

- ...

> The list could be longer in the future, but we don't want to make it too long here, hope you enjoy and let's get started.

Reviewer 2: ```suggestion
The list could be longer in the future, but we don't want to make it too long here. Hope you enjoy and let's get started.
```

- [Changelog](#changelog)
- [License](#license)

## Why

Many issues cannot be fixed easily without API changes. For example, see:

- <https://github.com/import-js/eslint-plugin-import/issues/1479>
- <https://github.com/import-js/eslint-plugin-import/issues/2108>
- <https://github.com/import-js/eslint-plugin-import/issues/2111>

> [`eslint-plugin-import`] refused to accept BREAKING CHANGES for these issues, so we had to fork it.

Author: > https://github.com/import-js/eslint-plugin-import/issues/2108#issuecomment-1205592491 -> https://github.com/import-js/eslint-plugin-import/issues/2108#issuecomment-1205864116: same thing but with a draft PR that has a pending question. So it's similarly that it's just difficult & more work to work with their constraints, right?

I don't get what's blocking the PR actually. All tests are passing. And when the maintainer says "is this potentially a breaking change", I just don't want to continue that PR.

All possible breaking changes are refused. All the issues are in the same condition IMO.

> [`eslint-plugin-import`] now claims in <https://github.com/un-ts/eslint-plugin-import-x/issues/170> that it will accept BREAKING CHANGES. However, still nothing is happening: <https://github.com/import-js/eslint-plugin-import/pull/3091>.

Author: > Have you checked with ljharb?

Why should I do that?

> [`eslint-plugin-import`] now claims in <https://github.com/un-ts/eslint-plugin-import-x/issues/170> that it will accept BREAKING CHANGES. However, still nothing is happening: <https://github.com/import-js/eslint-plugin-import/pull/3091>.

Reviewer 4: @JoshuaKGoldberg https://discord.com/channels/1227627367204257852/1300975143199309844

> [`eslint-plugin-import`] refuses to support the `exports` feature, and the maintainer even locked the feature request issue <https://github.com/import-js/eslint-plugin-import/issues/1810> to prevent future discussion. In the meantime, `eslint-plugin-import-x` now provides first-party support for the `exports` feature <https://github.com/un-ts/eslint-plugin-import-x/pull/209>, which will become the default in the next major version (v5).

Reviewer: > refuses

It looks to me like the root issue of fixing `resolve` (https://github.com/browserify/resolve/pull/224) is what was indicated as the desired fix, and that is similarly stalled due to difficulties in supporting the repo's long backwards compatibility requirements, yeah? So it's not that they're refusing, it's actively accepting - just difficult?

> ... to prevent future discussion

My read of https://github.com/import-js/eslint-plugin-import/issues/1810#issuecomment-1818010790 was that the lock was because one particular person was repeatedly arguming with the maintainer on that one day, and the conversation was getting heated & off-topic? The _"for a bit"_ makes me think the conversation was supposed to be un-locked and that just never happened, unintentionally. Maybe now is the time for a ping to re-open?

Author: > So it's not that they're refusing, it's actively accepting - just difficult?

Difficult due to backward compatibility, so it's just refusing.

> My read of https://github.com/import-js/eslint-plugin-import/issues/1810#issuecomment-1818010790 was that the lock was because one particular person was repeatedly arguming with the maintainer on that one day, and the conversation was getting heated & off-topic? The "for a bit" makes me think the conversation was supposed to be un-locked and that just never happened, unintentionally. Maybe now is the time for a ping to re-open?

You may help to reconstruct the sentence, it's just our feelings, everyone could have different feelings.

Reviewer 5: Difficulty due to constraints that I won't drop is not "refusing" - they have very different connotations in English.

Author: As I said, it's same for end users: refusing breaking changes.

Author: End users including me don't care about the technology background, we've already been using https://github.com/unrs/unrs-resolver.

Reviewer 5: The "end user" cares about being able to lint using resolution for their target platforms, not just the single node version they're running eslint with. That resolver has the same flaw as every single other alternative, which has been explained.

Reviewer 5: That's fine, but making slanderous comments about me (eg that i'm "refusing") isn't warranted or intellectually honest. "Clean" is also an opinion, to be clear, not a fact.

Author: > You may help to reconstruct the sentence, it's just our feelings, everyone could have different feelings.

We haven't resolved all the issues yet, but we are working on them, which could happen in the next major version (v5): <https://github.com/un-ts/eslint-plugin-import-x/issues/235>.

## Differences

So what are the differences from `eslint-plugin-import` exactly?

- we target [Node `^18.18.0 || ^20.9.0 || >=21.1.0`](https://github.com/un-ts/eslint-plugin-import-x/blob/8b2d6d3b612eb57fb68c3fddec25b02fc622df7c/package.json#L12) + [ESLint `^8.57.0 || ^9.0.0`](https://github.com/un-ts/eslint-plugin-import-x/blob/8b2d6d3b612eb57fb68c3fddec25b02fc622df7c/package.json#L71), while `eslint-plugin-import` targets [Node `>=4`](https://github.com/import-js/eslint-plugin-import/blob/da5f6ec13160cb288338db0c2a00c34b2d932f0d/package.json#L6) and [ESLint `^2 || ^3 || ^4 || ^5 || ^6 || ^7.2.0 || ^8 || ^9`](https://github.com/import-js/eslint-plugin-import/blob/da5f6ec13160cb288338db0c2a00c34b2d932f0d/package.json#L115C16-L115C64)

> - we don't depend on old and outdated dependencies, so [we have 49 dependencies](https://npmgraph.js.org/?q=eslint-plugin-import-x) compared to [117 dependencies for `eslint-plugin-import`](https://npmgraph.js.org/?q=eslint-plugin-import)

Reviewer: Data request: the counts and links to npmgraph are a good start, but don't immediately describe what the _real impact_ is. Dependency tree size is abstract and vague, and not an end-user metric (e.g. 1 of the 49 dependencies might be larger than all of the opposing 117 combined).

In order to get to any end-user-relevant info such as the actual unpacked module size you have to know to:

1. Open both links
2. Click on the root plugins in the visualizer
3. Find each _Module Size_ > _Unpacked Size (module + dependencies)_

I think most readers will not do that, so calling out the 6 MB vs 36 MB comparison here would be required for getting it known.

Author: We'll get even less dependencies when dropping `eslint-import-resolver-node`: https://github.com/un-ts/eslint-plugin-import-x/pull/272#discussion_r2024451326

> - `eslint-plugin-import` uses `tsconfig-paths` + `typescript` itself to load `tsconfig`s while we use the single `get-tsconfig` instead, which is much faster and cleaner

Reviewer: > much faster

Prove it! Show me the data! 😄

As a developer, I have been trained to not believe any performance comparison not backed up by data. One should always back perf comparisons up with repeatable experiments.

Note that isolated experiments in `get-tsconfig` itself are not enough. The package might be fast in isolation but equal -or even slower!- in real-world usage. I need actual data from this specific plugin use case.

Author: > Prove it! Show me the data! 😄

https://github.com/privatenumber/get-tsconfig#features

I can say for 3rd part libraries. And I'm not interested to prove it because it maybe hard as you said.

> - `eslint-plugin-import` uses `tsconfig-paths` + `typescript` itself to load `tsconfig`s while we use the single `get-tsconfig` instead, which is much faster and cleaner

Reviewer: > cleaner

Why would an end-user care about this?

Author: Because we're cleaning the ecosystem? https://e18e.dev/guide/cleanup.html

> - `eslint-plugin-import` uses [`resolve`] which doesn't support the `exports` field in `package.json` while we build our own rust-based resolver [`unrs-resolver`] instead, which is feature-rich and way more performant.

Reviewer: > feature-rich

This is abstract and non-tangible: what do you mean? How is this relevant to a reader?

Author: Feel free to PR to replace them.
